Allow to use existing storage stack

priteau opened the pull request:
This is part of task cedadev/jasmin-cluster-as-a-service#32.

priteau committed:
Pass NFS cluster info to Heat stack

Together with cedadev/jasmin-cluster-infra-ansible#17 this allows
storage information to be returned in the `info` output of the stack.

priteau committed:
Allow to use existing storage stack

If the variable storage_name is defined, get NFS storage information
from it. Requires stackhpc/ansible-role-cluster-nfs#4 to be merged
first.

oneswig commented on jasmin.yml:
```diff
@@ -7,6 +7,19 @@
+            stack_id: "{{ storage_name }}"
+        - name: Export storage information from stack outputs
+          set_fact:
+            storage_stack_info: "{{ stack_outputs['info'] | select('ne', '') | map('from_json') | selectattr('storage_type', '==', 'nfs') | first }}"
```
It would be good in a follow-on development to extract the storage server hostnames so that they can be added to `/etc/hosts` along with the nodes in the Slurm cluster.

Building on this it would also be good to parameterise the specification of NFS.

oneswig reviewed (approved):
Looking forward to trying this out, thanks Pierre.

priteau commented:
- [x] Test alternative approach of injecting hostvars
- [ ] @brtknr recommends using upstream `os_stack_facts` code: https://github.com/ansible/ansible/pull/47748
- [ ] Document what can be improved when upgrading to Queens

oneswig committed:
Merge branch 'master' into storage-stack

oneswig committed:
Add cluster_info data for NFS, BeeGFS and Gluster using group_vars
for the openstack group.

priteau committed:
Implement new approach for using existing storage

Based on information gathered from the stack outputs, we create a
storage server group with hosts and assign them required variables.

priteau committed:
Move the OpenStack auth config into extra vars

priteau committed:
Add missing code

brtkwr commented:
Nice thing about the upstream `os_stack_facts` is that it only requires `openstacksdk` package, it doesn't depend on `python-heatclient` which we are already having problems installing on a brand new AWX deployment due to lack of `gcc` package...

Typical output from upstream `os_stack_facts` module:
```
ok: [localhost] => {
    "openstack_stack": {
        "action": "UPDATE",
        "created_at": "2019-03-22T11:40:21",
        "creation_time": "2019-03-22T11:40:21",
        "deleted_at": null,
        "deletion_time": null,
        "description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "disable_rollback": true,
        "id": "bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
        "identifier": "k8s/bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
        "is_rollback_enabled": true,
        "location": {
            "cloud": "envvars",
            "project": {
                "domain_id": null,
                "domain_name": "jasmin",
                "id": "43d4c99f76d443a5a17b192e08fb2e30",
                "name": "caastest-U"
            },
            "region_name": "nova",
            "zone": null
        },
        "name": "k8s",
        "notification_topics": [],
        "outputs": [
            {
                "description": "No description given",
                "output_key": "info",
                "output_value": [
                    "",
                    ""
                ]
            },
            {
                "description": "No description given",
                "output_key": "node_groups",
                "output_value": [
                    {
                        "info": "",
                        "name": "kube-masters",
                        "nodes": [
                            {
                                "id": "bf07e06e-e6e4-486b-9fd0-aad716707507",
                                "ip": "192.171.139.249",
                                "name": "k8s-kube-masters-0"
                            }
                        ]
                    },
                    {
                        "info": "",
                        "name": "kube-workers",
                        "nodes": [
                            {
                                "id": "b6936242-6417-454f-91f1-ddb734fc8b77",
                                "ip": "192.168.3.176",
                                "name": "k8s-kube-workers-0"
                            },
                            {
                                "id": "7bda958f-b789-46cf-91a8-7a2fc008529b",
                                "ip": "192.168.3.175",
                                "name": "k8s-kube-workers-1"
                            },
                            {
                                "id": "da7e52af-b182-4a61-99f2-248530849554",
                                "ip": "192.168.3.178",
                                "name": "k8s-kube-workers-2"
                            }
                        ]
                    }
                ]
            },
            {
                "description": "No description given",
                "output_key": "deploy_private_key",
                "output_value": "..."
            }
        ],
        "owner": null,
        "parameters": {
            "OS::project_id": "43d4c99f76d443a5a17b192e08fb2e30",
            "OS::stack_id": "bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
            "OS::stack_name": "k8s",
            "cluster_group_count": "2",
            "cluster_groups": "[{u'nodenet_fips': [{u'ip': u'192.171.139.249', u'uuid': u'8723c186-d15f-44d5-9421-e12bf2a03325'}], u'name': u'kube-masters', u'root_volume_size': 0, u'num_nodes': 1, u'nodenet_resource': u'Cluster::NodeNet1WithPreallocatedFIP', u'user': u'centos', u'flavor': u'j3.medium', u'node_resource': u'Cluster::Node'}, {u'name': u'kube-workers', u'root_volume_size': 0, u'num_nodes': 3, u'nodenet_resource': u'Cluster::NodeNet1', u'user': u'centos', u'flavor': u'j3.medium', u'node_resource': u'Cluster::Node'}]",
            "cluster_image": "centos-7-20190104",
            "cluster_keypair": "brtknr-33236e18e6285585a7f58f1006f88349",
            "cluster_network": "caastest-U-internal"
        },
        "parent": null,
        "properties": {
            "capabilities": []
        },
        "stack_name": "k8s",
        "stack_owner": null,
        "stack_status": "UPDATE_COMPLETE",
        "stack_status_reason": "Stack UPDATE completed successfully",
        "stack_user_project_id": "33deb6ddc73a41c7912722695a941c40",
        "status": "COMPLETE",
        "status_reason": "Stack UPDATE completed successfully",
        "tags": null,
        "tempate_description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "template_description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "timeout_mins": 60,
        "updated_at": "2019-03-22T12:36:43",
        "updated_time": "2019-03-22T12:36:43"
    }
}
```

oneswig commented on tasks/storage_stack.yml:
```diff
@@ -0,0 +1,35 @@
+- name: Get stack outputs
```
This makes it look like new hosts in the inventory but would we ever be able to connect to them, given the storage stack will have its own private key?

priteau replied:
That's right, we cannot connect to them. We have to make sure we don't run tasks on them or gather facts, but we still use them for hostvars.

priteau committed:
Merge remote-tracking branch 'origin/master' into storage-stack

priteau committed:
Adapt slurm-infra to work with latest master

priteau committed:
Remove use of pre-allocated FIP

oneswig reviewed (approved):
Looks good, I can try building on this to add external BeeGFS in the same pattern.

priteau committed:
Pass nfs_export to the stackhpc.cluster-nfs role
